--- a/wp-fail2ban-bad-request-instant.inc.php   2017-06-10 15:10:33.205984255 +0000
+++ b/wp-fail2ban-bad-request-instant.inc.php   2017-10-02 10:30:25.526690211 +0000
@@ -246,7 +246,8 @@

         if ( defined( 'XMLRPC_REQUEST' ) && XMLRPC_REQUEST ) {
             $this->is_xmlrpc = true;
-        } elseif ( false !== strpos( $request_path, '/wp-login.php' ) ) {
+        // } elseif ( false !== strpos( $request_path, '/wp-login.php' ) ) {
+        } elseif ( false !== strpos( $request_path, '/downloader/' ) ) {
             $this->is_wplogin = true;
         } elseif ( false !== strpos( $request_path, $this->rest_url_prefix ) ) {
             $this->is_rest = true;
@@ -277,7 +278,7 @@
         // https://tools.ietf.org/html/rfc2616#section-10.4.15
         // 2500 bytes ~ deletion of 50 spam comments (GET form)
         if ( strlen( $_SERVER['REQUEST_URI'] ) > 2500 ) {
-            return 'bad_request_uri_length';
+// @MFIXME Cart item deletions produce very long AJAX URL-s            return 'bad_request_uri_length';
         }

         // Too big user agent
@@ -445,8 +445,6 @@
             && ( empty( $_SERVER['CONTENT_TYPE'] )
                 || ( $this->is_wplogin
                     && 0 !== stripos( $_SERVER['CONTENT_TYPE'], 'application/x-www-form-urlencoded' )
+                    && 0 !== stripos( $_SERVER['CONTENT_TYPE'], 'multipart/form-data' )
+                    && 0 !== stripos( $_SERVER['CONTENT_TYPE'], 'application/zip' )
                 )
                 || ( $this->is_xmlrpc
                     && 0 !== stripos( $_SERVER['CONTENT_TYPE'], 'text/xml' )
@@ -498,8 +499,10 @@
         }

         // Login request with non-empty username
-        if ( ! empty( $_POST['log'] ) ) {
-            $username = trim( $_POST['log'] );
+        // if ( ! empty( $_POST['log'] ) ) {
+        //     $username = trim( $_POST['log'] );
+        if ( ! empty( $_POST['login'] ) && ! empty( $_POST['login']['username'] ) ) {
+            $username = trim( $_POST['login']['username'] );

             // Banned usernames
             if ( in_array( strtolower( $username ), $this->names2ban ) ) {
@@ -565,7 +568,8 @@
         // WordPress test cookie
         if ( ! $this->allow_registration ) {
             if ( empty( $_SERVER['HTTP_COOKIE'] )
-                || false === strpos( $_SERVER['HTTP_COOKIE'], 'wordpress_test_cookie' )
+                // || false === strpos( $_SERVER['HTTP_COOKIE'], 'wordpress_test_cookie' )
+                || false === strpos( $_SERVER['HTTP_COOKIE'], 'adminhtml' )
             ) {
                 return 'bad_request_wplogin_test_cookie';
             }
@@ -586,7 +590,8 @@

         // Referer HTTP header
         if ( ! $this->allow_registration ) {
-            if ( false === strpos( parse_url( $referer, PHP_URL_PATH ), '/wp-login.php' ) ) {
+            // if ( false === strpos( parse_url( $referer, PHP_URL_PATH ), '/wp-login.php' ) ) {
+            if ( false === strpos( parse_url( $referer, PHP_URL_PATH ), '/downloader/' ) ) {
                 return 'bad_request_wplogin_referer_path';
             }
         }
